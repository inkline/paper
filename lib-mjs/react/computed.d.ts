import { Ref } from '@inkline/paper/types';
export declare function computed<T>(computeFn: () => T): Ref<T>;
