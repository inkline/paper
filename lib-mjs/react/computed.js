export function computed(computeFn) {
    return {
        get value() {
            return computeFn();
        }
    };
}
//# sourceMappingURL=computed.js.map