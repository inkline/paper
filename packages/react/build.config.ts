export default {
    entries: [
        {
            builder: 'mkdist',
            format: 'esm',
            ext: 'mjs',
            input: './src',
            declaration: true
        },
        {
            builder: 'mkdist',
            format: 'cjs',
            ext: 'cjs',
            input: './src',
            declaration: true
        }
    ]
};
