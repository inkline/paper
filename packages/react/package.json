{
  "name": "@inkline/paper-react",
  "description": "Inkline Paper React.js package.",
  "scripts": {
    "pre": "cp ../common/src/types.ts src/types.common.ts",
    "pretest": "npm run pre",
    "test": "vitest run",
    "prebuild": "npm run pre",
    "build": "tsc -d -p tsconfig.mjs.json && tsc -d -p tsconfig.cjs.json && mjscjs -p tsconfig.mjs.json && mjscjs -p tsconfig.cjs.json"
  },
  "devDependencies": {
    "@testing-library/react": "12.1.3",
    "@vitejs/plugin-react": "1.2.0",
    "jsdom": "19.0.0",
    "mjscjs": "1.0.9",
    "react": "17.0.2",
    "react-dom": "17.0.2",
    "rimraf": "3.0.2",
    "vite": "2.8.6",
    "vitest": "0.5.9"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/inkline/paper.git"
  },
  "license": "ISC"
}
