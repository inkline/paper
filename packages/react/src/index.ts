export * from './computed';
export * from './defineComponent';
export * from './definePlugin';
export * from './events';
export * from './fragment';
export * from './h';
export * from './helpers';
export * from './provide';
export * from './ref';
export * from './types';
export * from './watch';
