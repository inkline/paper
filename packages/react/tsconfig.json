{
    "extends": "../../tsconfig.json",
    "compilerOptions": {
        "types": [
            "@types/react"
        ]
    }
}
