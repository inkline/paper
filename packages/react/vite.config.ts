import { defineConfig } from 'vite';
import { resolve } from 'path';
import react from '@vitejs/plugin-react';

// https://vitejs.dev/config/
export default defineConfig({
    plugins: [
        react()
    ],
    resolve: {
        alias: [
            {
                find: /^@inkline\/paper\//,
                replacement: `${resolve(__dirname)}/src/`
            }
        ]
    },
    build: {
        lib: {
            entry: resolve(__dirname, 'src', 'index.ts'),
            name: 'InklinePaperReact',
            fileName: (format) => `react.${format}.js`
        },
        rollupOptions: {
            // make sure to externalize deps that shouldn't be bundled
            // into your library
            external: [
                'react'
            ],
            output: {
                exports: 'named',
                // Provide global variables to use in the UMD build
                // for externalized deps
                globals: {
                    react: 'React'
                }
            }
        }
    },
    test: {
        globals: true,
        environment: 'jsdom'
    }
});
