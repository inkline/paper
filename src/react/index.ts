export * from '@inkline/ucd/react/defineComponent';
export * from '@inkline/ucd/react/computed';
export * from '@inkline/ucd/react/h';
export * from '@inkline/ucd/react/helpers';
export * from '@inkline/ucd/react/ref';
export * from '@inkline/ucd/react/types';
