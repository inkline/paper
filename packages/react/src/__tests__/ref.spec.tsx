import React from 'react';
import { defineComponent, h, ref } from '../index';
import { describe, it, expect } from 'vitest';
import { fireEvent, render } from '@testing-library/react';

describe('react', () => {
    describe('ref()', () => {
        it('should create a reference value', () => {
            const Component = defineComponent({
                setup () {
                    const state = ref('hello');

                    expect(state).toEqual({
                        value: 'hello'
                    });

                    return { state };
                },
                render ({ state }) {
                    return h('div', {}, [state.value]);
                }
            });

            const wrapper = render(<Component /> as any);
            expect(wrapper.container.firstChild).toMatchSnapshot();
        });

        describe('getter', () => {
            it('should return current value', () => {
                const Component = defineComponent({
                    setup () {
                        const state = ref('hello');

                        expect(state.value).toEqual('hello');

                        return { state };
                    },
                    render ({ state }) {
                        return h('div', {}, [state.value]);
                    }
                });

                const wrapper = render(<Component /> as any);
                expect(wrapper.container.firstChild).toMatchSnapshot();
            });
        });

        describe('setter', () => {
            it('should set a new string value', async () => {
                const Component = defineComponent({
                    setup () {
                        const state = ref('hello');

                        const onClick = () => {
                            state.value = 'world';
                        };

                        return { state, onClick };
                    },
                    render ({ state, onClick }) {
                        return h('button', { onClick }, [state.value]);
                    }
                });

                const wrapper = render(<Component /> as any);
                await fireEvent.click(wrapper.container.firstChild as HTMLElement);

                expect(wrapper.container.firstChild).toMatchSnapshot();
                expect(wrapper.getByText('world')).toBeTruthy();
            });

            it('should set a new object value', async () => {
                const Component = defineComponent({
                    setup () {
                        const state = ref({ field: 'hello' });

                        const onClick = () => {
                            state.value = { field: 'world' };
                        };

                        return { state, onClick };
                    },
                    render ({ state, onClick }) {
                        return h('button', { onClick }, [state.value.field]);
                    }
                });

                const wrapper = render(<Component /> as any);
                await fireEvent.click(wrapper.container.firstChild as HTMLElement);

                expect(wrapper.container.firstChild).toMatchSnapshot();
                expect(wrapper.getByText('world')).toBeTruthy();
            });

            it('should set a new array value', async () => {
                const Component = defineComponent({
                    setup () {
                        const state = ref(['a']);

                        const onClick = () => {
                            state.value = state.value.concat(['b']);
                        };

                        return { state, onClick };
                    },
                    render ({ state, onClick }) {
                        return h('button', { onClick }, [state.value.join(', ')]);
                    }
                });

                const wrapper = render(<Component /> as any);
                await fireEvent.click(wrapper.container.firstChild as HTMLElement);

                expect(wrapper.container.firstChild).toMatchSnapshot();
                expect(wrapper.getByText('a, b')).toBeTruthy();
            });
        });
    });
});
