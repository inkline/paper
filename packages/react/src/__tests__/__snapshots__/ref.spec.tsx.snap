// Vitest Snapshot v1

exports[`react > ref() > getter > should return current value 1`] = `
<div>
  hello
</div>
`;

exports[`react > ref() > setter > should set a new array value 1`] = `
<button>
  a, b
</button>
`;

exports[`react > ref() > setter > should set a new object value 1`] = `
<button>
  world
</button>
`;

exports[`react > ref() > setter > should set a new string value 1`] = `
<button>
  world
</button>
`;

exports[`react > ref() > should create a reference value 1`] = `
<div>
  hello
</div>
`;
